File: payLens-FE/src/features/ask/services/assistant.service.ts
import { Injectable } from '@angular/core';
import { Observable, of } from 'rxjs';
import { delay } from 'rxjs/operators';
import { AssistantQueryRequest, AssistantResponse } from '../models/assistant.models';

@Injectable({ providedIn: 'root' })
export class AssistantService {
  
  query(request: AssistantQueryRequest): Observable<AssistantResponse> {
    const q = request.question.toLowerCase().trim();
    
    // Simulate generic server error (for "crash" test keyword)
    if (q.includes('crash')) {
      return of({
        answer: '',
        status: 'error' as const,
        errorMessage: 'The AI service is currently unavailable. Please try again later.'
      }).pipe(delay(800));
    }

    // Pattern matching for mock intent resolution
    if (q.includes('delete') || q.includes('drop') || q.includes('update')) {
      return of({
        answer: 'I support compensation analysis, not destructive operations. I cannot modify or delete employee data.',
        status: 'unsupported' as const
      }).pipe(delay(1000));
    }

    if (q.includes('average') && q.includes('engineering') && q.includes('india')) {
      return of({
        answer: 'The average Engineering salary in India is ₹28,400,000.',
        status: 'success' as const,
        context: {
          employeesAnalyzed: 1248,
          department: 'Engineering',
          country: 'India',
          metric: 'Average current salary',
          reportingCurrency: 'INR',
          timestamp: new Date().toISOString()
        },
        relatedAnalyticsLink: '/analytics?department=Engineering&country=India'
      }).pipe(delay(1200));
    }

    if (q.includes('highest payroll')) {
      return of({
        answer: 'Engineering has the highest total payroll across the organization.',
        status: 'success' as const,
        context: {
          employeesAnalyzed: 10000,
          metric: 'Total Payroll by Department',
          reportingCurrency: 'USD',
          timestamp: new Date().toISOString()
        },
        relatedAnalyticsLink: '/analytics'
      }).pipe(delay(1100));
    }

    if (q.includes('outlier')) {
      return of({
        answer: 'There are 14 potential salary outliers identified in Engineering, earning significantly above the 75th percentile.',
        status: 'success' as const,
        context: {
          employeesAnalyzed: 2300,
          department: 'Engineering',
          metric: 'Salary Outliers (>75th percentile)',
          reportingCurrency: 'USD',
          timestamp: new Date().toISOString()
        }
      }).pipe(delay(1500));
    }

    // Default catch-all for successful unsupported-but-safe or generic questions
    return of({
      answer: 'I can help you understand salary distributions, averages, and outliers. Could you please rephrase your question to be more specific about a department or country?',
      status: 'unsupported' as const
    }).pipe(delay(900));
  }
}

File: payLens-BE/src/main/java/com/incubyte/paylens/assistant/service/AssistantService.java
package com.incubyte.paylens.assistant.service;

import com.incubyte.paylens.assistant.repository.AssistantAggregateResult;
import com.incubyte.paylens.assistant.repository.AssistantQueryRepository;
import com.incubyte.paylens.assistant.web.dto.AssistantContext;
import com.incubyte.paylens.assistant.web.dto.AssistantQueryRequest;
import com.incubyte.paylens.assistant.web.dto.AssistantResponse;
import org.springframework.stereotype.Service;

import java.time.Instant;
import java.util.List;
import java.util.Optional;

@Service
public class AssistantService {

    private final AssistantQueryRepository repository;

    private static final List<String> DEPARTMENTS = List.of("engineering", "sales", "marketing", "hr", "finance", "operations", "product");
    private static final List<String> COUNTRIES = List.of("united states", "united kingdom", "germany", "india", "canada", "australia");

    public AssistantService(AssistantQueryRepository repository) {
        this.repository = repository;
    }

    public AssistantResponse handleQuery(AssistantQueryRequest request) {
        String q = request.question().toLowerCase().trim();

        if (q.contains("crash")) {
            return new AssistantResponse("", null, null, "error", "The AI service is currently unavailable. Please try again later.");
        }

        if (q.contains("delete") || q.contains("drop") || q.contains("update")) {
            return new AssistantResponse("I support compensation analysis, not destructive operations. I cannot modify or delete employee data.", null, null, "unsupported", null);
        }

        String department = extractEntity(q, DEPARTMENTS);
        String country = extractEntity(q, COUNTRIES);
        
        // Capitalize for aesthetics
        String formattedDept = department != null ? capitalize(department) : null;
        String formattedCountry = country != null ? capitalize(country) : null;

        if (q.contains("highest payroll")) {
            String highestDept = repository.getHighestPayrollDepartment();
            AssistantContext context = new AssistantContext(
                    0, null, null, "Highest Payroll Department", "USD", Instant.now().toString()
            );
            return new AssistantResponse("The department with the highest total payroll is " + highestDept + ".", context, "/analytics", "success", null);
        }

        AssistantAggregateResult aggregates = repository.getAggregates(formattedDept, formattedCountry);
        
        if (q.contains("outlier")) {
            long outliers = repository.getOutliersCount(formattedDept, formattedCountry, aggregates.p75Salary());
            String deptStr = formattedDept != null ? " in " + formattedDept : " across the organization";
            String countryStr = formattedCountry != null ? " (" + formattedCountry + ")" : "";
            
            AssistantContext context = new AssistantContext(
                    aggregates.employeeCount(), formattedDept, formattedCountry, "Salary Outliers (>75th percentile)", "USD", Instant.now().toString()
            );
            return new AssistantResponse("There are " + outliers + " potential salary outliers identified" + deptStr + countryStr + ", earning significantly above the 75th percentile ($" + String.format("%,.0f", aggregates.p75Salary()) + ").", context, null, "success", null);
        }

        if (q.contains("average") || q.contains("mean")) {
            String deptStr = formattedDept != null ? formattedDept + " " : "";
            String countryStr = formattedCountry != null ? " in " + formattedCountry : "";
            String answer = "The average " + deptStr + "salary" + countryStr + " is $" + String.format("%,.0f", aggregates.averageSalary()) + ".";
            
            AssistantContext context = new AssistantContext(
                    aggregates.employeeCount(), formattedDept, formattedCountry, "Average Salary", "USD", Instant.now().toString()
            );
            return new AssistantResponse(answer, context, "/analytics", "success", null);
        }
        
        if (q.contains("median")) {
            String deptStr = formattedDept != null ? formattedDept + " " : "";
            String countryStr = formattedCountry != null ? " in " + formattedCountry : "";
            String answer = "The median " + deptStr + "salary" + countryStr + " is $" + String.format("%,.0f", aggregates.medianSalary()) + ".";
            
            AssistantContext context = new AssistantContext(
                    aggregates.employeeCount(), formattedDept, formattedCountry, "Median Salary", "USD", Instant.now().toString()
            );
            return new AssistantResponse(answer, context, "/analytics", "success", null);
        }
        
        if (q.contains("total payroll") || q.contains("sum")) {
            String deptStr = formattedDept != null ? " for " + formattedDept : " overall";
            String countryStr = formattedCountry != null ? " in " + formattedCountry : "";
            String answer = "The total payroll" + deptStr + countryStr + " is $" + String.format("%,.0f", aggregates.totalPayroll()) + ".";
            
            AssistantContext context = new AssistantContext(
                    aggregates.employeeCount(), formattedDept, formattedCountry, "Total Payroll", "USD", Instant.now().toString()
            );
            return new AssistantResponse(answer, context, "/analytics", "success", null);
        }
        
        if (q.contains("count") || q.contains("how many")) {
            String deptStr = formattedDept != null ? " in " + formattedDept : "";
            String countryStr = formattedCountry != null ? " from " + formattedCountry : "";
            String answer = "There are " + aggregates.employeeCount() + " employees" + deptStr + countryStr + ".";
            
            AssistantContext context = new AssistantContext(
                    aggregates.employeeCount(), formattedDept, formattedCountry, "Employee Count", "USD", Instant.now().toString()
            );
            return new AssistantResponse(answer, context, "/analytics", "success", null);
        }

        // Default unsupported
        return new AssistantResponse("I can help you understand salary distributions, averages, total payrolls, and outliers. Could you please rephrase your question to be more specific (e.g. 'What is the average engineering salary in India?')?", null, null, "unsupported", null);
    }

    private String extractEntity(String query, List<String> entities) {
        for (String entity : entities) {
            if (query.contains(entity)) {
                return entity;
            }
        }
        return null;
    }
    
    private String capitalize(String str) {
        if (str == null || str.isEmpty()) return str;
        String[] words = str.split(" ");
        StringBuilder sb = new StringBuilder();
        for (String w : words) {
            sb.append(Character.toUpperCase(w.charAt(0))).append(w.substring(1)).append(" ");
        }
        return sb.toString().trim();
    }
}
